src/main/java/optimal_sort/MergeSort.java

```diff
@@ -0,0 +1,54 @@
+package optimal_sort;
+
+import sort.Sorter;
+import java.util.ArrayList;
+import java.util.List;
+
+public class MergeSort implements Sorter {
+
+    private static void merge(final List<Integer> nums, final List<Integer> leftList, final List<Integer> rightList, final int left, final int right) {
+
+        int i = 0;
+        int j = 0;
+        int k = 0;
+        while (i < left && j < right) {
+            if (leftList.get(i) <= rightList.get(j)) {
+                nums.set(k++, leftList.get(i++));
+            }
+            else {
+                nums.set(k++, rightList.get(j++));
+            }
+        }
+        while (i < left) {
+            nums.set(k++, leftList.get(i++));
+        }
+        while (j < right) {
+            nums.set(k++, rightList.get(j++));
+        }
+    }
+
+    @Override
+    public void sort(final List<Integer> nums) {
+        final int n = nums.size();
+        if (n < 2) {
+            return;
+        }
+
+        final int mid = n / 2;
+        final List<Integer> leftList = new ArrayList<>();
+        final List<Integer> rightList = new ArrayList<>();
+
+        for (int i = 0; i < mid; i++) {
+            final int value = nums.get(i);
+            leftList.add(value);
+        }
+        for (int i = mid; i < n; i++) {
+            final int value = nums.get(i);
+            rightList.add(value);
+        }
+        sort(leftList);
+        sort(rightList);
+
+        merge(nums, leftList, rightList, mid, n - mid);
+    }
+}
```

Comment:
In my opinion you should keep sort method simple, but merge more complicated. Example, how I did it
```
private void mergeSort(final int[] nums, final int left, final  int right) {
        if (left >= right) {
            return;
        }
        final int mid = left + (right - left) / 2;
        mergeSort(nums, left, mid);
        mergeSort(nums, mid + 1, right);
        merge(nums, left, right, mid);
    }

    private void merge(final int[] nums, final int left, final  int right, final int mid ) {
        final int[] copy = new int[right - left + 1];
        int i = left;
        int j = mid + 1;
        int k = 0;
        while (i <= mid && j <= right) {
            if (nums[i] <= nums[j]) {
                copy[k] = nums[i];
                i++;
            } else {
                copy[k] = nums[j];
                j++;
            }
            k++;
        }
        while (i <= mid) {
            copy[k] = nums[i];
            i++;
            k++;
        }
        while (j <= right) {
            copy[k] = nums[j];
            j++;
            k++;
        }
        i = left;
        for (final int value : copy) {
            nums[i] = value;
            i++;
        }
    }
```